### PR TITLE
Rest api

### PR DESCRIPTION
**Pull request by StephenDsouza90**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,102 @@
...
+# Trustly Parliament Challenge
+
+### Problems
+1. Not being able to merger speeches data with members data
+
+speeches
+```
+[
+  {'anforande_id': [], 'dok_datum': [], 'talare': [], 'parti': [], 'protokoll_url_www': [], 'dok_titel': [], 'intressent_id': []}, 
+  {'anforande_id': [], 'dok_datum': [], 'talare': [], 'parti': [], 'protokoll_url_www': [], 'dok_titel': [], 'intressent_id': []},
```
Comment (dsouzajude) [inline README.md]: Howcome these are empty lists?

File: README.md [hunk 2]
```diff
@@ -1 +1,102 @@
...
+{'valkrets': [], 'bild_url_192': [], 'uppgift': [], 'intressent_id_check': []}, 
+{'valkrets': [], 'bild_url_192': [], 'uppgift': [], 'intressent_id_check': []}, 
+{'valkrets': [], 'bild_url_192': [], 'uppgift': [], 'intressent_id_check': []}, 
+{'valkrets': [], 'bild_url_192': [], 'uppgift': [], 'intressent_id_check': []}, 
+{'valkrets': [], 'bild_url_192': [], 'uppgift': [], 'intressent_id_check': []}, 
+{'valkrets': [], 'bild_url_192': [], 'uppgift': [], 'intressent_id_check': []}, 
+{'valkrets': [], 'bild_url_192': [], 'uppgift': [], 'intressent_id_check': []}, 
+{'valkrets': [], 'bild_url_192': [], 'uppgift': [], 'intressent_id_check': []}, 
+{'valkrets': [], 'bild_url_192': [], 'uppgift': [], 'intressent_id_check': []}, 
+{'valkrets': [], 'bild_url_192': [], 'uppgift': [], 'intressent_id_check': []},
```
Comment (dsouzajude) [inline README.md]: These are all the same, looks like duplicate values. Are you sure this is the data you get from the members api?

File: README.md [hunk 3]
```diff
@@ -1 +1,102 @@
-Test
\ No newline at end of file
+# Trustly Parliament Challenge
+
+### Problems
+1. Not being able to merger speeches data with members data
+
+speeches
```
Comment (dsouzajude) [inline README.md]: You don't need to give so many examples, for simplicity, just show 2 or max 3 results from the speeches api

File: README.md [hunk 4]
```diff
@@ -1 +1,102 @@
...
+  
+  {'anforande_id': ['42102c9a-e312-ea11-912c-901b0e9b71a8'], 'dok_datum': ['2019-11-29'], 'talare': ['Infrastrukturministern Tomas Eneroth (S)'], 'parti': ['S'], 'protokoll_url_www': ['http://www.riksdagen.se/sv/Dokument-Lagar/Kammaren/Protokoll/Riksdagens-snabbprotokoll_H70941/#anf163'], 'dok_titel': ['Protokoll 2019/20:41 Fredagen den 29 november'], 'intressent_id': ['0284192765516']}, 
+  
+  {'anforande_id': ['41102c9a-e312-ea11-912c-901b0e9b71a8'], 'dok_datum': ['2019-11-29'], 'talare': ['Jens Holm (V)'], 'parti': ['V'], 'protokoll_url_www': ['http://www.riksdagen.se/sv/Dokument-Lagar/Kammaren/Protokoll/Riksdagens-snabbprotokoll_H70941/#anf162'], 'dok_titel': ['Protokoll 2019/20:41 Fredagen den 29 november'], 'intressent_id': ['0216534495014']}, 
+  
+  {'anforande_id': ['40102c9a-e312-ea11-912c-901b0e9b71a8'], 'dok_datum': ['2019-11-29'], 'talare': ['Infrastrukturministern Tomas Eneroth (S)'], 'parti': ['S'], 'protokoll_url_www': ['http://www.riksdagen.se/sv/Dokument-Lagar/Kammaren/Protokoll/Riksdagens-snabbprotokoll_H70941/#anf161'], 'dok_titel': ['Protokoll 2019/20:41 Fredagen den 29 november'], 'intressent_id': ['0284192765516']}, 
+```
+
+members
```
Comment (dsouzajude) [inline README.md]: You don't need to give so many examples, for simplicity, just show 2 or max 3 results from the members api

File: README.md [hunk 5]
```diff
@@ -1 +1,102 @@
...
+{'valkrets': ['Stockholms kommun'], 'bild_url_192': ['http://data.riksdagen.se/filarkiv/bilder/ledamot/979e9852-69ed-4077-8140-52ab99414a6c_192.jpg'], 'uppgift': [], 'intressent_id_check': ['014386615025']}, 
+
+{'valkrets': ['Stockholms kommun'], 'bild_url_192': ['http://data.riksdagen.se/filarkiv/bilder/ledamot/5b7ac02c-0819-426d-bc5d-41b8c21ca4dd_192.jpg'], 'uppgift': [], 'intressent_id_check': ['0216534495014']}
+]
+```
+
+## Approach
+>> Speech data already provides the list of latest ten speeches.
+
+>> Merger the relevant members data to the speeches data.
```
Comment (dsouzajude) [inline README.md]: What do you mean by `Merger`? Do you mean `Merge` the relevant members ... ? English doesn't make sense.

File: README.md [hunk 6]
```diff
@@ -1 +1,102 @@
...
+
+>> Match foreign key which is "intressent_id" .. in english stakeholders_id
```
Comment (dsouzajude) [inline README.md]: Why do you need to match this? 

Always provide a "Why" for every statement. This shows your logic and explanation towards solving a problem.

File: README.md [hunk 7]
```diff
@@ -1 +1,102 @@
...
+
+>> Present relevant combined data.
```
Comment (dsouzajude) [inline README.md]: Give an example of relevant combined data ..

File: README.md [hunk 8]
```diff
@@ -1 +1,102 @@
-Test
\ No newline at end of file
+# Trustly Parliament Challenge
+
+### Problems
+1. Not being able to merger speeches data with members data
```
Comment (dsouzajude) [inline README.md]: Whatever problem you face, always try to solve the problem on a small scale. So basically what i mean by this is, take two dicts and try to merge them. Once you know how to do it, try to apply the same logic and solution on a big scale for the real problem.

File: README.md (hunk 1 above)
Comment (StephenDsouza90) [reply README.md]: problem debugged.

File: README.md (hunk 2 above)
Comment (StephenDsouza90) [reply README.md]: problem debugged.

File: README.md (hunk 7 above)
Comment (StephenDsouza90) [reply README.md]: example given

File: README.md (hunk 6 above)
Comment (StephenDsouza90) [reply README.md]: reason given

File: README.md (hunk 5 above)
Comment (StephenDsouza90) [reply README.md]: change made

File: README.md [hunk 9]
```diff
@@ -1 +1,141 @@
-Test
\ No newline at end of file
+# Trustly Parliament Challenge
+
+## Introduction
```
Comment (dsouzajude) [inline README.md]: In the Introduction discuss what the problem is and what you need to do, then go about  explaining your approach to solving the problem. Also put a link to the Trustly problem challenge when describing the task.

File: README.md [hunk 10]
```diff
@@ -1 +1,141 @@
...
+
+The parliament challenge by Trustly is approached as follows:
+
+1. Get speeches resource from its respective apis. Since speeches data is in a nested dict, the for loop has been used to iterate over the required dict which is "anforande". The speeches api allows for extracting the ten latest speeches.
```
Comment (dsouzajude) [inline README.md]: Is it really nested? How?

File: README.md [hunk 11]
```diff
@@ -1 +1,141 @@
...
+
+> Speech api: http://data.riksdagen.se/anforandelista/?anftyp=Nej&sz=10&utformat=json
+
+2. Get members resource from its respective apis. Since members data is in a nested dict, the for loop has been used to iterate over the required dict which is "person". All members data has been extracted.
```
Comment (dsouzajude) [inline README.md]: Is it really nested? How?

File: README.md [hunk 12]
```diff
@@ -1 +1,141 @@
...
+  
+  {'valkrets': ['Stockholms kommun'], 'bild_url_192': ['http://data.riksdagen.se/filarkiv/bilder/ledamot/979e9852-69ed-4077-8140-52ab99414a6c_192.jpg'], 'uppgift': [], 'intressent_id_FK': ['014386615025']}, 
+  
+  {'valkrets': ['Stockholms kommun'], 'bild_url_192': ['http://data.riksdagen.se/filarkiv/bilder/ledamot/5b7ac02c-0819-426d-bc5d-41b8c21ca4dd_192.jpg'], 'uppgift': [], 'intressent_id_FK': ['0216534495014']}
+  ]
+```
+
+### Required output
+
+when common object matches
```
Comment (dsouzajude) [inline README.md]: What do you mean by `common object matches`? What's the issue?

Comment (StephenDsouza90) [reply README.md]: The common object between both speeches and members is "intressent_id". I am using this to make a match between the speeches dict to members dict.

File: README.md (hunk 10 above)
Comment (StephenDsouza90) [reply README.md]: the resource is as follows:

{
  "anforandelista": {
    "@antal": "10",
    "anforande": [
      {"keys": "values",
       "keys": "values"
}]

so it becomes nested

File: README.md (hunk 11 above)
Comment (StephenDsouza90) [reply README.md]: the resource is as follows:

{
  "personlista": {
    "@systemdatum": "2019-12-02 16:51:13",
    "@hits": "349",
    "person": [
      {{"keys": "values",
        "keys": "values"
}]

so it becomes nested

File: README.md [hunk 13]
```diff
@@ -1 +1,124 @@
...
+- The debate subject (avsnittsrubrik)
+- A link to the speech (HATEOAS style) (protokoll_url_www)
+
+The link to the challenge can be found on https://github.com/trustly/parliament-challenge
+
+## Approach
+
+The challenge is approached as follows:
+
+1. The function `get_speeches()` is responsible for getting the speeches resources from its respective apis. The speeches api has a filter to extract the ten latest speeches. Since the data is in a nested dict, the for loop has been used to iterate over to the required dict which is "anforande". Speech api: http://data.riksdagen.se/anforandelista/?anftyp=Nej&sz=10&utformat=json
```
Comment (dsouzajude) [inline README.md]: Because it's in a nested dict, it doesn't mean you need to use the for loop.

File: README.md [hunk 14]
```diff
@@ -1 +1,124 @@
...
+
+2. The function `get_members()` is responsible for getting the members resources from its respective apis. Here all members data has been extracted. Since the data is in a nested dict, the for loop has been used to iterate over to the required dict which is "person". Members api: http://data.riksdagen.se/personlista/?iid=&fnamn=&enamn=&f_ar=&kn=&parti=&valkrets=&rdlstatus=&org=&utformat=json&sort=sorteringsnamn&sortorder=asc&termlista=
```
Comment (dsouzajude) [inline README.md]: Because it's in a nested dict, it doesn't mean you need to use the for loop.

File: README.md [hunk 15]
```diff
@@ -1 +1,124 @@
...
+
+3. Now since the ten latest speeches are in the `get_speeches()` function, a list of dict has been created in the `create_speeches_dict()` function contianing only the relevant keys and values of speeches. An additional key and value `intressent_id` has been added which can be used as a foreign key and be useful for merging the speeches data to members data. The relevant keys are:
```
Comment (dsouzajude) [inline README.md]: `foreign key` is only a terminology used in data bases. You can instead use the word "reference".

File: README.md [hunk 16]
```diff
@@ -1 +1,124 @@
...
+
+- A unique ID, representing the speech itself (anforande_id)
+- Date of speech (dok_datum)
+- The name of the speaker, and only the name (tilltalsnamn)
+- Political affiliation (parti)
+- The official e-mail address (uppgift)
+- Constituency (valkrets)
+- A decent sized image that could be used in a web site listing (bild_url_192)
+- The debate subject (avsnittsrubrik)
+- A link to the speech (HATEOAS style) (protokoll_url_www)
```
Comment (dsouzajude) [inline README.md]: Research on what HATEOAS means and how to format this link via HATEOAS style.

File: README.md (hunk 9 above)
Comment (StephenDsouza90) [reply README.md]: added link and discussed problem

File: README.md (hunk 13 above)
Comment (StephenDsouza90) [reply README.md]: looped through relevant key to get speeches keys and values.

File: README.md (hunk 14 above)
Comment (StephenDsouza90) [reply README.md]: looped through relevant key to get members keys and values.

File: README.md (hunk 15 above)
Comment (StephenDsouza90) [reply README.md]: used reference id instead

File: README.md (hunk 16 above)
Comment (StephenDsouza90) [reply README.md]: added HATEOAS style

File: README.md [hunk 17]
```diff
@@ -1 +1,110 @@
...
```
Comment (dsouzajude) [inline README.md]: For each field in the bracket, also indicate which api it's coming from, i.e. members api or speeches api.

File: README.md [hunk 18]
```diff
@@ -1 +1,110 @@
...
+
+The link to the challenge can be found on https://github.com/trustly/parliament-challenge
+
+## Approach
+
+The challenge is approached as follows:
```
Comment (dsouzajude) [inline README.md]: Revise the entire description. Don't go into very technical detailed implementations. Just give a very simple description to your approach on what the api is about and how you can use it. Don't include any function names in your code. Try to just be very high level and simple. Example:

The speech api is .... It contains the following fields .. These fields represent ....
The members api is ..  It contains the following fields .. These fields represent ...

Since the resources from the speech api has member related info, we can use that field to get details of the member using the member api .. 

Once we get that, we can merge it like ... (give an example).

Example request / response from the speech api (omit all fields, show only the ones you're interested in) and just show max two responses and do a `...` after the list to indicate there are more results:
>> ... curl ....
```json
{
}
```


Example request / response from the member api:
>> ... curl ....
```json
{
}
```

Then show an example of the merged result.
```json
{ ...}
```

Comment (StephenDsouza90) [reply README.md]: revised approach

File: README.md (hunk 17 above)
Comment (StephenDsouza90) [reply README.md]: added fields

File: parliament_challenge.py [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+    domain = 'http://data.riksdagen.se'
+    speeches = 'anforandelista'
+    size = 10
+    format_type = 'json'
+    response = requests.get('{}/{}/?anftyp=Nej&sz={}&utformat={}'.format(domain, speeches, size, format_type))
+    if response.status_code == 200:
+        data = response.json()
+        for k, v in data.items():
+            anforande = v["anforande"]
+            return anforande
```
Comment (dsouzajude) [inline parliament_challenge.py]: This doesn't make sense. It's a bug. Debug and see what is returned.!

File: parliament_challenge.py [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+
+def create_speeches_dict():
+    """
+    Use the get_speeches_api() function to create a dict of the relevant items.
+    """
+
+    speeches_data = get_speeches_api()
+    speeches_list = []
+    for speeches in speeches_data:
+        anforande_id = [v for k, v in speeches.items() if k == "anforande_id"]
```
Comment (dsouzajude) [inline parliament_challenge.py]: Why do you need a for loop here to get the value of `anforande_id`?

Comment (dsouzajude) [reply parliament_challenge.py]: Why you're making it so complicated? Make it simple

Comment (dsouzajude) [reply parliament_challenge.py]: The same for all keys! Think again how to make it simpler. All you're supposed to do is get the value for a key in a given dict and you know the key. so think again how to do it.

**Comment by dsouzajude** [comment]
The logic is too complicated. Revise it and make it simpler!

File: parliament_challenge.py (hunk 19 above)
Comment (StephenDsouza90) [reply parliament_challenge.py]: updated

File: parliament_challenge.py (hunk 20 above)
Comment (StephenDsouza90) [reply parliament_challenge.py]: updated

File: README.md [hunk 21]
```diff
@@ -1 +1,85 @@
...
+
+### Member's API
+
+The member's api represents the details of the members of the Swedish parliament. In order get the relevant member data and link the speech to a member, a reference key **intressent_id** (which is common in both apis) is used as a filter option. The relevant keys and values that this api has are: **tilltalsnamn, valkrets, bild_url_192**.
+
+Link to the member's api: http://data.riksdagen.se/personlista/?iid=&utformat=json
+
+### Merge
+
+In order to merge a speech to a member, the relevant keys and values are filtered from the the speech api and the reference key **intressent_id** from a speech is passed as an arguement and parameter to get the relevant (filtered) member data which is then updated with that relevant speech data and stored in a list of dict.
```
Comment (dsouzajude) [inline README.md]: This definition is too detailed. Make it simpler. Describe it in plain english without any technical implementation details. Explain it like as if you're explaining it to a non-technical person.

File: parliament_challenge.py [hunk 22]
```diff
@@ -0,0 +1,111 @@
+import requests
+import waitress
+import flask
+from flask import Flask, request, json
+
+
+def filter_speeches_dict(speeches):
+    """
+    Filter speeches data for relevant items.
```
Comment (dsouzajude) [inline parliament_challenge.py]: This is not correct. You're filtering speech data for the relevant "keys", not items.

And also say that the relevant keys are `anforande_id`, `dok_datum` .. and so on ..

File: parliament_challenge.py [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+                    }],
+            "intressent_id": s.get("intressent_id")
+            }
+        speeches_list.append(speechesDict)
+    return speeches_list
+
+
+def filter_member_dict(member):
+    """
+    Filter member data for relevant items.
```
Comment (dsouzajude) [inline parliament_challenge.py]: This is not correct. You're filtering speech data for the relevant "keys", not items.

And also say that the relevant keys are `tilltalsnamn`, `valkrets` .. and so on ..

File: parliament_challenge.py [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+        "valkrets": member.get("valkrets"),
+        "bild_url_192": member.get("bild_url_192"),
+        "intressent_id": member.get("intressent_id")
+    }
+    return membersDict
+
+
+def get_member_data(intressent_id):
+    """
+    Get memeber data filter for relevant items.
```
Comment (dsouzajude) [inline parliament_challenge.py]: Elaborate a little more, the definition in the comment doesn't make sense.

File: parliament_challenge.py [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+    """
+
+    domain = 'http://data.riksdagen.se'
+    members = 'personlista'
+    format_type = 'json'
+    response = requests.get('{}/{}/?iid={}&utformat={}'.format(domain, members, intressent_id, format_type))
+    if response.status_code == 200:
```
Comment (dsouzajude) [inline parliament_challenge.py:66]: What about the else part? Never forget the else part.

File: parliament_challenge.py [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+        data = response.json()
+        member = data["personlista"]["person"]
+        filtered_member = filter_member_dict(member)
+        return filtered_member
+
+
+def get_speeches_data(anftyp, size):
+    """
+    Get speeches data and filter for relevant items.
```
Comment (dsouzajude) [inline parliament_challenge.py]: What is relevant items?

File: parliament_challenge.py [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+    Loop through filtered speeches to get "intressent id" 
+    and pass it as an argurment to the get_member_data 
+    to get member data.
```
Comment (dsouzajude) [inline parliament_challenge.py]: Simplify the explanation. Just say .. 

.... 
For each speech we query the member api to get the member data and then merge it with the speech data.
...

Comment (dsouzajude) [reply parliament_challenge.py]: You don't have to say which functions you're calling and what arguments you pass to it. The programmer or anyone reading your code will understand it from reading the code. The description should only be very high level .. in plain english what the function is supposed to do, what arguments it takes and what it returns.

File: README.md (hunk 21 above)
Comment (dsouzajude) [reply README.md]: You don't have to say which functions you're calling and what arguments you pass to it. The programmer or anyone reading your code will understand it from reading the code. The description should only be very high level .. in plain english what the function is supposed to do, what arguments it takes and what it returns.

File: parliament_challenge.py [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+        
+
+def create_app():
+    app = Flask("Parliament Challenge")
+
+    @app.route('/latest-speeches', methods=['GET'])
+    def get_latest_speeches():
+        """
+        GET request:
+            curl -H "Content-Type: application/json" -X GET -d "{\"anftyp\":\"Nej\", \"size\":10}" "localhost:8080/latest-speeches"
```
Comment (dsouzajude) [inline parliament_challenge.py]: The `-d` flag refers to sending "data" and the client only sends data in `PUT` and `POST` calls, not `GET`.

For the `GET` REST calls, use the  query parameters. Google on query parameters using flask for REST APIs or see the previous projects how you passed query parameters and processed them!

File: README.md (hunk 21 above)
Comment (StephenDsouza90) [reply README.md]: updated explanation: 

In order to merge a speech to a member, the function will filter the respective apis for the relevant keys and values and link a speech to a member through the reference key **intressent_id**. The function will then update the speech data with member data by combining both data sets together and present a merged data.

File: parliament_challenge.py (hunk 22 above)
Comment (StephenDsouza90) [reply parliament_challenge.py]: updated:

    Filtering speeches data for relevant keys.
    Relevant keys are anforande_id, dok_datum,
    parti, avsnittsrubrik, links and intressent_id.

File: parliament_challenge.py (hunk 23 above)
Comment (StephenDsouza90) [reply parliament_challenge.py]: updated:
    Filtering member data for relevant keys.
    Relevant keys are tilltalsnamn, valkrets,
    bild_url_192 and intressent_id.

File: parliament_challenge.py (hunk 24 above)
Comment (StephenDsouza90) [reply parliament_challenge.py]: updated:

    Get a member data from the member's api
    and filter for relevant keys.

File: parliament_challenge.py (hunk 25 above)
Comment (StephenDsouza90) [reply parliament_challenge.py:66]: return []

File: parliament_challenge.py (hunk 26 above)
Comment (StephenDsouza90) [reply parliament_challenge.py]: updated

File: parliament_challenge.py (hunk 27 above)
Comment (StephenDsouza90) [reply parliament_challenge.py]: updated

File: parliament_challenge.py (hunk 28 above)
Comment (StephenDsouza90) [reply parliament_challenge.py]: used query parameter.
curl request
>> curl -X GET "localhost:8080/latest-speeches/?anftyp=Nej&sz=2"

to get values from url
anftyp = flask.request.args["anftyp"]
size = flask.request.args["sz"]

File: parliament_challenge.py [hunk 29]
```diff
@@ -0,0 +1,119 @@
...
+    members = 'personlista'
+    format_type = 'json'
+    response = requests.get('{}/{}/?iid={}&utformat={}'.format(domain, members, intressent_id, format_type))
+    if response.status_code == 200:
+        data = response.json()
+        member = data["personlista"]["person"]
+        filtered_member = filter_member_dict(member)
+        return filtered_member
+    else:
+        return []
```
Comment (dsouzajude) [inline parliament_challenge.py]: Why are you returning a list?

File: parliament_challenge.py [hunk 30]
```diff
@@ -0,0 +1,119 @@
...
+    domain = 'http://data.riksdagen.se'
+    speeches = 'anforandelista'
+    format_type = 'json'
+    response = requests.get('{}/{}/?anftyp={}&sz={}&utformat={}'.format(domain, speeches, anftyp, size, format_type))
+    if response.status_code == 200:
+        data = response.json()
+        speeches = data["anforandelista"]["anforande"]
+        filtered_speeches = filter_speeches_dict(speeches)
+        for speech in filtered_speeches:
+            member_data = get_member_data(speech["intressent_id"])
```
Comment (dsouzajude) [inline parliament_challenge.py]: What if the `speech["intressent_id"]` is null?

Comment (StephenDsouza90) [reply parliament_challenge.py]: updated:

    if response.status_code == 200:
        data = response.json()
        if data["personlista"]["@hits"] == "1":
            member = data["personlista"]["person"]
            filtered_member = filter_member_dict(member)
            return filtered_member
        elif data["personlista"]["@hits"] == "0":
            nullMemberDict = {"member data": "null"}
            return nullMemberDict
    elif response.status_code == 404:
         return not_found
    else:
        return failed

File: README.md [hunk 31]
```diff
@@ -1 +1,86 @@
...
+
+In order to merge a speech to a member, the function will filter the respective apis for the relevant keys and values and link a speech to a member through the reference key **intressent_id**. The function will then update the speech data with member data by combining both data sets together and present a merged data.
+
+## Server
+
+The `create app()` is responsible for creating the server and the `main()` is used to run the server. This application uses a localhost and port 8080.
+
+## How to run locally
+
+This app uses the curl request **curl -H "Content-Type: application/json" -X GET -d "{\"anftyp\":\"Nej\", \"size\":10}" "localhost:8080/latest-speeches"** which has been mapped to the `get_latest_speeches()` in the `create_app()`.
```
Comment (dsouzajude) [inline README.md]: The `-d` flag is not used, right? You're using query params?

File: parliament_challenge.py [hunk 32]
```diff
@@ -0,0 +1,140 @@
...
+
+def filter_member_dict(member):
+    """
+    Filtering member data for relevant keys.
+    Relevant keys are tilltalsnamn, valkrets,
+    bild_url_192, uppgift and intressent_id.
+    """
+
+    if member["personuppgift"] is None:
+        email = "null"
```
Comment (dsouzajude) [inline parliament_challenge.py]: Set email = None, not string "null". The json dumps will automatically convert it to null.

File: parliament_challenge.py [hunk 33]
```diff
@@ -0,0 +1,140 @@
...
+    and filter for relevant keys.
+    """
+
+    domain = 'http://data.riksdagen.se'
+    members = 'personlista'
+    format_type = 'json'
+    response = requests.get('{}/{}/?iid={}&utformat={}'.format(domain, members, intressent_id, format_type))
+    if response.status_code == 200:
+        data = response.json()
+        if data["personlista"]["@hits"] == "1":
```
Comment (dsouzajude) [inline parliament_challenge.py:69]: Add some line comments with the `#` here, it's good to comment what the logic is doing when it looks a bit complex.

File: parliament_challenge.py [hunk 34]
```diff
@@ -0,0 +1,140 @@
...
+    then merge it with the speech data.
+    """
+
+    domain = 'http://data.riksdagen.se'
+    speeches = 'anforandelista'
+    format_type = 'json'
+    response = requests.get('{}/{}/?anftyp={}&sz={}&utformat={}'.format(domain, speeches, anftyp, size, format_type))
+    if response.status_code == 200:
+        data = response.json()
+        if data["anforandelista"]["@antal"] > "1":
```
Comment (dsouzajude) [inline parliament_challenge.py:98]: Add some line  comments here

File: parliament_challenge.py [hunk 35]
```diff
@@ -0,0 +1,140 @@
...
+            filtered_speeches = filter_speeches_dict(speeches)
+            for speech in filtered_speeches:
+                member_data, code = get_member_data(speech["intressent_id"])        
+                speech.update(member_data)
+            return filtered_speeches, code
+        elif data["anforandelista"]["@antal"] == "1":
+            lst = []
+            speeches = data["anforandelista"]["anforande"]
+            lst.append(speeches)
+            filtered_speeches = filter_speeches_dict(lst)
```
Comment (dsouzajude) [inline parliament_challenge.py]: You can also do something more simple:

```python


            # When @antal is <= 1, only a single dictionary is returned, not a list
            # So we simulate a list by putting the speech inside a list with `[]`.
            speeches = [
                data["anforandelista"]["anforande"]
            ]
            filtered_speeches = filter_speeches_dict(speeches)

```

File: parliament_challenge.py [hunk 36]
```diff
@@ -0,0 +1,149 @@
...
+        if data["personlista"]["@hits"] == "1":
+            member = data["personlista"]["person"]
+            filtered_member = filter_member_dict(member)
+            return filtered_member, response.status_code
+        # In case a member detail is non-existent,
+        # returning an empty dict to indicate that 
+        # the detail does not exist in the member data.
+        elif data["personlista"]["@hits"] == "0":
+            nullMemberDict = {}
+            return nullMemberDict, response.status_code
```
Comment (dsouzajude) [inline parliament_challenge.py]: These two lines can be reduced to one line:

```python
return {}, response.status_code
```

File: parliament_challenge.py [hunk 37]
```diff
@@ -0,0 +1,149 @@
...
+    then merge it with the speech data.
+    """
+
+    domain = 'http://data.riksdagen.se'
+    speeches = 'anforandelista'
+    format_type = 'json'
+    response = requests.get('{}/{}/?anftyp={}&sz={}&utformat={}'.format(domain, speeches, anftyp, size, format_type))
+    if response.status_code == 200:
+        data = response.json()
+        # When a client requests for more than one speech.
```
Comment (dsouzajude) [inline parliament_challenge.py]: This comment is wrong right? The client doesn't request for more than one speech? The speech api  server returned more than one speech.

File: parliament_challenge.py [hunk 38]
```diff
@@ -0,0 +1,149 @@
...
+        if data["anforandelista"]["@antal"] > "1":
+            speeches = data["anforandelista"]["anforande"]
+            filtered_speeches = filter_speeches_dict(speeches)
+            for speech in filtered_speeches:
+                member_data, code = get_member_data(speech["intressent_id"])        
+                speech.update(member_data)
+            return filtered_speeches, code
+        # When a client requests only for one speech.
+        # As a dict is returned, that speech is stored 
+        # in a list and then filtered.
```
Comment (dsouzajude) [inline parliament_challenge.py]: Same here, the client is not requesting for one speech! Revise the comment, doesn't make sense.

Comment (dsouzajude) [reply parliament_challenge.py]: Just indicate in very simple words what the logic is about in the case when the speech api returns only 1 speech. You know that it returns a dict and not a list, so we need to handle this situation and adapt it to the `filter_speeches_dict` which accepts a list instead of a dict, so we convert this into a list to be accepted by the `filter_speeches_dict` function.

**Review by dsouzajude — APPROVED** [review]
Looks good!
